WalletWasabi.Backend/Controllers/WabiSabi/IdempotencyRequestCache.cs

```diff
@@ -0,0 +1,107 @@
+
+		/// <summary>Timeout specifying how long a request response can stay in memory.</summary>
+		private static TimeSpan CacheTimeout { get; } = TimeSpan.FromMinutes(5);
+
+		private static TimeSpan RequestTimeout { get; } = TimeSpan.FromMinutes(1);
+
+		public IdempotencyRequestCache(IMemoryCache cache)
+		{
+			ResponseCache = cache;
+		}
+
+		/// <summary>Guards <see cref="ResponseCache"/>.</summary>
+		/// <remarks>Unfortunately, <see cref="CacheExtensions.GetOrCreate{TItem}(IMemoryCache, object, Func{ICacheEntry, TItem})"/> is not atomic.</remarks>
+		/// <seealso href="https://github.com/dotnet/runtime/issues/36499"/>
+		private object ResponseCacheLock { get; } = new();
+
+		/// <remarks>Guarded by <see cref="ResponseCacheLock"/>.</remarks>
+		private IMemoryCache ResponseCache { get; }
+
+		/// <typeparam name="TRequest">
+		/// <see langword="record"/>s are preferred as <see cref="object.GetHashCode"/>
+		/// and <see cref="object.Equals(object?)"/> are generated for <see langword="record"/> types automatically.
+		/// </typeparam>
+		/// <typeparam name="TResponse">Type associated with <typeparamref name="TRequest"/>. The correspondence should be 1:1 mapping.</typeparam>
+		public async Task<TResponse> GetCachedResponseAsync<TRequest, TResponse>(TRequest request, ProcessRequestDelegateAsync<TRequest, TResponse> action, CancellationToken cancellationToken = default)
+			where TRequest : notnull
+			where TResponse : class
+		{
+			bool callAction = false;
+			TaskCompletionSource<TResponse> responseTcs;
+
+			while (true)
+			{
+
+				lock (ResponseCacheLock)
+				{
+					if (!ResponseCache.TryGetValue(request, out responseTcs))
+					{
+						callAction = true;
+						responseTcs = new();
+						ResponseCache.Set(request, responseTcs, DateTimeOffset.UtcNow.Add(CacheTimeout));
+					}
+				}
+
+				TResponse result;
+
+				if (callAction)
+				{
+					try
+					{
+						result = await action(request, cancellationToken).ConfigureAwait(false);
+						responseTcs.SetResult(result);
+						return result;
+					}
+					catch (Exception e)
+					{
+						responseTcs.SetException(e);
+						throw;
+					}
+				}
+				else
+				{
+					using CancellationTokenSource timeoutCts = new(RequestTimeout);
+					using var linkedCts = CancellationTokenSource.CreateLinkedTokenSource(timeoutCts.Token, cancellationToken);
+
+					try
+					{
+						result = await responseTcs.Task.WithAwaitCancellationAsync(linkedCts.Token).ConfigureAwait(false);
+						return result;
+					}
+					catch (OperationCanceledException e)
+					{
+						if (e.CancellationToken == cancellationToken)
+						{
+							// Cancelled by application shutting down or when the HTTP request is cancelled.
+							throw;
+						}
+					}
+					catch (Exception e) when (e is WabiSabiProtocolException or WabiSabiCryptoException)
+					{
+						// 
+						throw;
+					}
+					catch (Exception)
+					{
+						lock (ResponseCacheLock)
+						{
+							ResponseCache.Remove(request);
+						}
```

Comment:
If I'm reading this right, this is the only branch that would result in the loop being restarted. I think you can just return an error response in this circumstance, and eliminate the loop making it easier to follow the control flow.

Reply:
there is actually one more possibility and that is L83:

```cs
				catch (OperationCanceledException e)  // <-- L83
					{
						if (e.CancellationToken == cancellationToken)
						{
							// Cancelled by application shutting down or when the HTTP request is cancelled.
							throw;
						}
					}
```

when a first request times out, we repeat at the moment for the second one. 

I remember that you mentioned that we should not have timeouts for requests which would mean to get rid of `RequestTimeout` really. Is that correct? Should we remove it?

Reply:
So the current behavior is that: *When a first request is canceled, then nothing is stored in cache. A second request will attempt to do the same work as the first one again because there is no entry in the cache.* So it is meant to be robust and it can create more delays but in practice I would not expect it to happen often unless somebody does that on purpose but then modifying a request is easier anyway so it does not look like a potential security risk.

So I believe I implement the second option you talk about.